Fix UB from misalignment and provenance widening in `std::sys::windows`

thomcc opened the pull request:
This fixes two types of UB:

1. Reading past the end of a reference in types like `&c::REPARSE_DATA_BUFFER` (see https://github.com/rust-lang/unsafe-code-guidelines/issues/256). This is fixed by using `addr_of!`. I think there are probably a couple more cases where we do this for other structures, and will look into it in a bit.

2. Failing to ensure that a `[u8; N]` on the stack is sufficiently aligned to convert to a `REPARSE_DATA_BUFFER`. ~~This was done by introducing a new `AlignedAs` struct that allows aligning one type to the alignment of another type. I expect there are other places where we have this issue too, or I wouldn't introduce this type, but will get to them after this lands.~~

    ~~Worth noting, it *is* implemented in a way that can cause problems depending on how we fix #81996, but this would be caught by the test I added (and presumably if we decide to fix that in a way that would break this code, we'd also introduce a `#[repr(simple)]` or `#[repr(linear)]` as a replacement for this usage of `#[repr(C)]`).~~
    
    Edit: None of that is still in the code, I just went with a `Align8` since that's all we'll need for almost everything we want to call.

These are more or less "potential UB" since it's likely at the moment everything works fine, although the alignment not causing issues might just be down to luck (and x86 being forgiving).

~~NB: I've only ensured this check builds, but will run tests soon.~~ All tests pass, including stage2 compiler tests.

r? @ChrisDenton

thomcc commented on library/std/src/sys/windows/fs.rs:
```diff
@@ -47,6 +47,9 @@ pub struct ReadDir {
 }
 
+type AlignedReparseBuf =
+    AlignedAs<c::REPARSE_DATA_BUFFER, [u8; c::MAXIMUM_REPARSE_DATA_BUFFER_SIZE]>;
```
I'll do a followup that makes this use `MaybeUninit<u8>`. I was going to do that first, but then noticed that this code was wrong.

ChrisDenton commented on library/std/src/sys/windows/c.rs:
```diff
@@ -511,6 +513,8 @@ pub struct REPARSE_DATA_BUFFER {
     pub rest: (),
```
Does this really need to end with a zst? Given the issues, it might be worth exploring other options. The usual way to do this in msvc is to end with `[u8; 1]` (or similar). Or it could be a union of supported reparse types. Or it could be both, which is the [canonical definition](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/ns-ntifs-_reparse_data_buffer).

An alternative is to (ab)use the fact that the `REPARSE_DATA_BUFFER` header is 64bits on all supported platforms. Which is convenient for aligning the data (and is likely no accident).

thomcc replied:
I think it's not great for it to end in a ZST, yeah.

thomcc replied:
This wouldn't really change the situation with UB that much though -- we still couldn't return a `&REPARSE_DATA_BUFFER` without narrowing the provenance such that it'd be UB to access `PathBuffer`.

ChrisDenton replied:
Yeah, that's true. I'm not minded to block this PR on removing the zst but I do think we should be aiming to do so, and in general leaning toward not using them in Windows ffi so as to side-step #81996 completely. Especially as the Windows API doesn't really use them.

thomcc replied:
I think that one would be fine, the issue isn't with all ZSTs, so much as zero-sized arrays like `[T; 0]` (since in a struct `T foo[0];` is the same as `T *foo;` on MSVC).

ChrisDenton commented on library/std/src/sys/windows/c.rs:
```diff
@@ -511,6 +513,8 @@ pub struct REPARSE_DATA_BUFFER {
     pub rest: (),
 }
 
+/// NB: Use carefully! In general using this as a reference is likely to get the
+/// provenance wrong for the `PathBuffer` field!
```
This is fine but it is a shame we need to keep repeating variations of this comment. But eh, it's fine. Encapsulating the unsafely is what the higher level APIs are for.

thomcc replied:
Beats someone missing it -- by repeating it, it shows up in auto-complete in rust-analyzer.

ChrisDenton commented on library/std/src/sys/windows/mod.rs:
```diff
@@ -329,3 +329,26 @@ pub fn abort_internal() -> ! {
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct AlignedAs<Aligner, Alignee: ?Sized> {
+    /// Use `[Aligner; 0]` as a sort of `PhantomAlignNextField<Aligner>`. This
+    /// is a bit of a hack, and could break (in a way that's caught by tests) if
+    /// #81996 is fixed.
+    aligner: [Aligner; 0],
```
I'm not especially enthusiastic about more hacks but it is at least more principled than what came before, and the tests help ensure it works. So... it's fine.

ChrisDenton commented on library/std/src/sys/windows/fs.rs:
```diff
@@ -389,7 +392,7 @@ impl File {
             attr.file_size = info.AllocationSize as u64;
             attr.number_of_links = Some(info.NumberOfLinks);
             if attr.file_type().is_reparse_point() {
-                let mut b = [0; c::MAXIMUM_REPARSE_DATA_BUFFER_SIZE];
+                let mut b = AlignedReparseBuf::new([0; c::MAXIMUM_REPARSE_DATA_BUFFER_SIZE]);
                 if let Ok((_, buf)) = self.reparse_point(&mut b) {
                     attr.reparse_tag = buf.ReparseTag;
                 }
```
Hm, you might want to update this so it matches the (non-UWP) code above?

thomcc replied:
Nice catch. UWP passes a check build now, at least.

ChrisDenton reviewed (commented):
Ok, I think this is basically fine. Only one thing that I think needs to be addressed, the other comments are just me being resigned to Rust not having better support for this atm. 🤷

ChrisDenton commented:
Oh, just noticed, I think you missed a spot:

https://github.com/rust-lang/rust/blob/1b8025a24c4b063d2566671f0664e5dfc263c2a4/library/std/src/sys/windows/fs.rs#L1348

Though this is in `symlink_junction` which is only used for tests..

thomcc commented:
> EDIT: Wait no that's a different buffer... for some reason.

It's still wrong the way it's written, I think?

That said there are a couple places in this file that are like this that I didn't touch (for example, it's not 100% clear to me what the best fix for the DirBuff stuff would be, but I was just going to tackle it later).

ChrisDenton commented:
> It's still wrong the way it's written, I think?

Yeah, but I'm ok with this PR being more focused. That said, I'm also more than happy for you to fix that too!

thomcc commented:
I fixed it. It actually was wrong in another way, since it was turning a pointer derived from `data.as_ptr()` into a mutable pointer, which is wrong because it's mutating a pointer derived from `&self`, which can't give out mutable pointers.

(Tragically `as_mut_ptr()` would also be wrong -- that goes through `&mut self` which would invalidate the other pointers derived from it, so I just added a new local for `data.value.as_mut_ptr()`)

ChrisDenton commented:
lgtm

@bors r+

bors commented:
:pushpin: Commit d9c760db43d8ab701a71f633d820efc72d8cedea has been approved by `ChrisDenton`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=ChrisDenton d9c760db43d8ab701a71f633d820efc72d8cedea -->
<!-- homu: {"type":"Approved","sha":"d9c760db43d8ab701a71f633d820efc72d8cedea","approver":"ChrisDenton","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

thomcc commented:
@bors r-

After a quick discussion with @ChrisDenton I'm actually going to simplify this a bit.

thomcc commented:
I'll remove the `donotland` CI changes after it passes (don't have a convenient windows box right now)

ChrisDenton commented on library/std/src/sys/windows/fs.rs:
```diff
@@ -696,13 +705,21 @@ impl<'a> Iterator for DirBuffIter<'a> {
+            assert!(buffer.len() >= header_sz);
             let info = buffer.as_ptr().cast::<c::FILE_ID_BOTH_DIR_INFO>();
-            let next_entry = (*info).NextEntryOffset as usize;
+            assert!(info.is_aligned());
+            let len_bytes = (*info).FileNameLength as usize;
+            // Note: we don't need `header_sz - 2` here, since the `[WCHAR; 1]`
+            // in the header and the trailing nul char (which always present but
+            // not counted in `len_bytes`) cancel each-other out.
+            assert!(buffer.len() >= header_sz + len_bytes);
```
I think some of these asserts are a touch paranoid because we should be able to assume the OS is acting sensibly when filling the buffer. However, I don't think they do any harm either and it is possible they could catch programmer error or problems with miri's emulation.

thomcc replied:
These checks are too annoying to write correctly without `offset_of!`, so I removed the ones not related to alignment.

rust-log-analyzer commented:
The job **`x86_64-msvc-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/8089103913?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/52ebbecd0e28ce680ba6cf4642329eabb907875e/checks/8089103913/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Updating files:  98% (35222/35940)
Updating files:  99% (35581/35940)
Updating files: 100% (35940/35940)
Updating files: 100% (35940/35940), done.
Note: switching to 'refs/remotes/pull/101171/merge'.
You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

---
  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at 058c38ad Merge 52ebbecd0e28ce680ba6cf4642329eabb907875e into a0d07093f80a0206f42d3dbada66212eda52b694
[command]"C:\Program Files\Git\bin\git.exe" log -1 --format='%H'
'058c38addf08d3559ca5a93a3c9032cb3d509ff3'
'058c38addf08d3559ca5a93a3c9032cb3d509ff3'
##[group]Run echo "[CI_PR_NUMBER=$num]"
echo "[CI_PR_NUMBER=$num]"
env:
  CI_JOB_NAME: x86_64-msvc-1
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
failures:

---- [incremental] src/test\incremental\lto.rs stdout ----

error in revision `rpass2`: compilation failed!
status: exit code: 101
command: PATH="D:\a\rust\rust\build\x86_64-pc-windows-msvc\stage2\bin;C:\Program Files (x86)\Windows Kits\10\bin\x64;C:\Program Files (x86)\Windows Kits\10\bin\10.0.22000.0\x64;C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64;C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64;D:\a\rust\rust\build\x86_64-pc-windows-msvc\stage0-bootstrap-tools\x86_64-pc-windows-msvc\release\deps;D:\a\rust\rust\build\x86_64-pc-windows-msvc\stage0\bin;D:\a\rust\rust\ninja;D:\a\rust\rust\msys2\mingw64\bin;C:\hostedtoolcache\windows\Python\3.10.6\x64\Scripts;C:\hostedtoolcache\windows\Python\3.10.6\x64;C:\msys64\usr\bin;D:\a\rust\rust\sccache;C:\Program Files\MongoDB\Server\5.0\bin;C:\aliyun-cli;C:\vcpkg;C:\cf-cli;C:\Program Files (x86)\NSIS;C:\tools\zstd;C:\Program Files\Mercurial;C:\hostedtoolcache\windows\stack\2.7.5\x64;C:\cabal\bin;C:\ghcup\bin;C:\tools\ghc-9.4.1\bin;C:\Program Files\dotnet;C:\mysql\bin;C:\Program Files\R\R-4.2.1\bin\x64;C:\SeleniumWebDrivers\GeckoDriver;C:\Program Files (x86)\sbt\bin;C:\Program Files (x86)\GitHub CLI;C:\Program Files\Git\bin;C:\Program Files (x86)\pipx_bin;C:\npm\prefix;C:\hostedtoolcache\windows\go\1.17.13\x64\bin;C:\hostedtoolcache\windows\Python\3.7.9\x64\Scripts;C:\hostedtoolcache\windows\Python\3.7.9\x64;C:\hostedtoolcache\windows\Ruby\2.5.9\x64\bin;C:\tools\kotlinc\bin;C:\hostedtoolcache\windows\Java_Temurin-Hotspot_jdk\8.0.345-1\x64\bin;C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\wbin;C:\ProgramData\kind;C:\Program Files\Eclipse Foundation\jdk-8.0.302.8-hotspot\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;C:\Windows\System32\OpenSSH;C:\ProgramData\Chocolatey\bin;C:\Program Files\Docker;C:\Program Files\PowerShell\7;C:\Program Files\Microsoft\Web Platform Installer;C:\Program Files\dotnet;C:\Program Files\Microsoft SQL Server\130\Tools\Binn;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit;C:\Program Files (x86)\Microsoft SQL Server\110\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\120\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\130\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\140\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\150\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\160\DTS\Binn;C:\Program Files\OpenSSL\bin;C:\Strawberry\c\bin;C:\Strawberry\perl\site\bin;C:\Strawberry\perl\bin;C:\ProgramData\chocolatey\lib\pulumi\tools\Pulumi\bin;C:\Program Files\TortoiseSVN\bin;C:\Program Files\CMake\bin;C:\ProgramData\chocolatey\lib\maven\apache-maven-3.8.6\bin;C:\Program Files\Microsoft Service Fabric\bin\Fabric\Fabric.Code;C:\Program Files\Microsoft SDKs\Service Fabric\Tools\ServiceFabricLocalClusterManager;C:\Program Files\nodejs;C:\Program Files\Git\cmd;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\bin;C:\Program Files\GitHub CLI;C:\tools\php;C:\Program Files (x86)\sbt\bin;C:\SeleniumWebDrivers\ChromeDriver;C:\SeleniumWebDrivers\EdgeDriver;C:\Program Files\Amazon\AWSCLIV2;C:\Program Files\Amazon\SessionManagerPlugin\bin;C:\Program Files\Amazon\AWSSAMCLI\bin;C:\Program Files (x86)\Google\Cloud SDK\google-cloud-sdk\bin;C:\Program Files (x86)\Microsoft BizTalk Server;C:\Program Files\LLVM\bin;C:\Users\runneradmin\.dotnet\tools;C:\Users\runneradmin\.cargo\bin;C:\Users\runneradmin\AppData\Local\Microsoft\WindowsApps" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage2\\bin\\rustc.exe" "D:\\a\\rust\\rust\\src/test\\incremental\\lto.rs" "-Zthreads=1" "--target=x86_64-pc-windows-msvc" "--cfg" "rpass2" "-C" "incremental=D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\test\\incremental\\lto\\lto.inc" "-Z" "incremental-verify-ich" "-O" "--error-format" "json" "--json" "future-incompat" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-o" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\test\\incremental\\lto\\a.exe" "-Crpath" "-Cdebuginfo=0" "-Lnative=D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\native\\rust-test-helpers" "-C" "lto" "-L" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\test\\incremental\\lto\\auxiliary"
stdout: none
Some tests failed in compiletest suite=incremental mode=incremental host=x86_64-pc-windows-msvc target=x86_64-pc-windows-msvc
--- stderr -------------------------------
thread 'rustc' panicked at 'expected 160 >= 164', library\std\src\sys\windows\fs.rs:716:13
   0:     0x7ffa9ac2ebbe - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h80e537c8d117ac29
   1:     0x7ffa9ac77e64 - core::fmt::write::h697008b405f58a60
   1:     0x7ffa9ac77e64 - core::fmt::write::h697008b405f58a60
   2:     0x7ffa9ac21a39 - <std::io::IoSlice as core::fmt::Debug>::fmt::h5561555eef88c4be
   3:     0x7ffa9ac326bf - std::panicking::default_hook::h3ae960ddf6d782b7
   4:     0x7ffa9ac322c2 - std::panicking::default_hook::h3ae960ddf6d782b7
   5:     0x7ffa8f1a5e50 - rustc_driver[9fe596517a89631]::handle_options
   6:     0x7ffa9ac32f29 - std::panicking::rust_panic_with_hook::hb567257b9600e961
   7:     0x7ffa9ac32cdd - <std::panicking::begin_panic_handler::StrPanicPayload as core::panic::BoxMeUp>::get::h4b063cdc7988f0db
   8:     0x7ffa9ac2f85f - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h80e537c8d117ac29
   9:     0x7ffa9ac32960 - rust_begin_unwind
  10:     0x7ffa9acadfe5 - core::panicking::panic_fmt::hcf568bd960affb10
  11:     0x7ffa9ac378b3 - <std::sys::windows::fs::DirBuffIter as core::iter::traits::iterator::Iterator>::next::h502d727d26883dd2
  12:     0x7ffa9ac38e9f - std::sys::windows::fs::remove_dir_all::ha643638f1b9a4913
  13:     0x7ffa9ac38b58 - std::sys::windows::fs::remove_dir_all::ha643638f1b9a4913
  14:     0x7ffa9418d9a1 - rustc_incremental[8b4db11f44a70fc5]::persist::fs::garbage_collect_session_directories
  15:     0x7ffa941889ae - rustc_incremental[8b4db11f44a70fc5]::persist::fs::finalize_session_directory
  16:     0x7ffa8f35c0c0 - <rustc_interface[ca2fb27880a37302]::queries::Linker>::link
  17:     0x7ffa8f18c0cb - RINvCsiNrYEkoyjgO_10rustc_span15with_source_mapINtNtCsixhsLQS10J6_4core6result6ResultuNtCskH3Qvge38XG_12rustc_errors15ErrorGuaranteedENCINvNtCshmerJrE6cWk_15rustc_interface9interface23create_compiler_and_runBJ_NCNvCsRc7gVEVAXZ_12rustc_driver12run_compiler
  18:     0x7ffa8f1aaff7 - rustc_interface[ca2fb27880a37302]::interface::create_compiler_and_run::<core[d7e8dadaf58ecf86]::result::Result<(), rustc_errors[f10ace7f3e14d91e]::ErrorGuaranteed>, rustc_driver[9fe596517a89631]::run_compiler::{closure#1}>
  19:     0x7ffa8f1851d7 - RINvMs_Cs3oRijBCYv8M_10scoped_tlsINtB5_9ScopedKeyNtCsiNrYEkoyjgO_10rustc_span14SessionGlobalsE3setNCINvNtCshmerJrE6cWk_15rustc_interface9interface12run_compilerINtNtCsixhsLQS10J6_4core6result6ResultuNtCskH3Qvge38XG_12rustc_errors15ErrorGuaranteedENCNvCsRc
  20:     0x7ffa8f266a76 - RINvNtNtCsdhAVEcL1GmI_3std10sys_common9backtrace28___rust_begin_short_backtraceNCINvNtCshmerJrE6cWk_15rustc_interface4util31run_in_thread_pool_with_globalsNCINvNtB1m_9interface12run_compilerINtNtCsixhsLQS10J6_4core6result6ResultuNtCskH3Qvge38XG_12rustc_er
  21:     0x7ffa8f269d07 - RNCINvNvXsh_NtNtNtCsixhsLQS10J6_4core4iter8adapters7flattenINtBa_13FlattenCompatppENtNtNtBe_6traits8iterator8Iterator8try_fold7flattenINtNtNtBg_5slice4iter4IterNtNtCsiNrYEkoyjgO_10rustc_span13span_encoding4SpanEuINtNtNtBg_3ops12control_flow11ControlFlowTN
  22:     0x7ffa9ac4896c - std::sys::windows::thread::Thread::new::h2fb59e7b309f0bc0
  23:     0x7ffacb207974 - BaseThreadInitThunk
  24:     0x7ffaccc1a2f1 - RtlUserThreadStart
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.


note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.65.0-nightly (058c38add 2022-08-30) running on x86_64-pc-windows-msvc

note: compiler flags: -Z threads=1 -C incremental -Z incremental -Z ui-testing -Z deduplicate-diagnostics=no -C rpath -C debuginfo=0 -C lto
query stack during panic:
end of query stack
------------------------------------------

---

test result: FAILED. 153 passed; 1 failed; 3 ignored; 0 measured; 0 filtered out; finished in 48.34s

Build completed unsuccessfully in 0:29:51
make: *** [Makefile:73: ci-subset-1] Error 1
```

</details>

thomcc commented:
Windows passes (https://github.com/rust-lang/rust/runs/8091513764?check_suite_focus=true, https://github.com/rust-lang/rust/runs/8091513602?check_suite_focus=true), so I've undone the CI commit.

ChrisDenton commented:
@bors r+

bors commented:
:pushpin: Commit c41f21b3e4e992d5c5c11dae880dc81192a12653 has been approved by `ChrisDenton`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=ChrisDenton c41f21b3e4e992d5c5c11dae880dc81192a12653 -->
<!-- homu: {"type":"Approved","sha":"c41f21b3e4e992d5c5c11dae880dc81192a12653","approver":"ChrisDenton","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
